src/Features/Core/Portable/TaskList/AbstractTaskListIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,123 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.TodoComments;
+
+namespace Microsoft.CodeAnalysis.TaskList
+{
+    internal abstract partial class AbstractTaskListIncrementalAnalyzer : IncrementalAnalyzerBase
+    {
+        private readonly object _gate = new();
+        private ImmutableArray<string> _lastTokenList = ImmutableArray<string>.Empty;
+        private ImmutableArray<TaskListItemDescriptor> _lastDescriptors = ImmutableArray<TaskListItemDescriptor>.Empty;
+
+        /// <summary>
+        /// Set of documents that we have reported an non-empty set of todo comments for.  Used so that we don't bother
+        /// notifying the host about documents with empty-todo lists (the common case). Note: no locking is needed for
+        /// this set as the incremental analyzer is guaranteed to make all calls sequentially to us.
+        /// </summary>
+        private readonly HashSet<DocumentId> _documentsWithItems = new();
+
+        protected abstract ValueTask ReportTaskListItemsAsync(DocumentId documentId, ImmutableArray<TaskListItem> data, CancellationToken cancellationToken);
+        protected abstract ValueTask<TaskListOptions> GetOptionsAsync(CancellationToken cancellationToken);
+
+        public override Task RemoveDocumentAsync(DocumentId documentId, CancellationToken cancellationToken)
+        {
+            // Remove the doc id from what we're tracking to prevent unbounded growth in the set.
+
+            // If the doc that is being removed is not in the set of docs we've told the host has todo comments,
+            // then no need to notify the host at all about it.
+            if (!_documentsWithItems.Remove(documentId))
+                return Task.CompletedTask;
+
+            // Otherwise, report that there should now be no todo comments for this doc.
+            return ReportTaskListItemsAsync(documentId, ImmutableArray<TaskListItem>.Empty, cancellationToken).AsTask();
+        }
+
+        private ImmutableArray<TaskListItemDescriptor> GetTaskListItemDescriptors(ImmutableArray<string> tokenList)
+        {
+            lock (_gate)
+            {
+                if (!tokenList.SequenceEqual(_lastTokenList))
+                {
+                    _lastDescriptors = TaskListItemDescriptor.Parse(tokenList);
+                    _lastTokenList = tokenList;
+                }
+
+                return _lastDescriptors;
+            }
+        }
+
+        private static ITaskListService? GetTaskListService(Document document)
+        {
+            // Legacy compat until TypeScript moves to EA pattern.
+            var todoService = document.GetLanguageService<ITodoCommentService>();
+            if (todoService != null)
+                return new TodoCommentServiceWrapper(todoService);
```

Comment:
this is the EA transition point.  we defer to the old api if present, but will gracefully fall over to the new one if it is not.